cEP-0017.md

```diff
@@ -0,0 +1,366 @@
+   it can be installed from `PyPi` using `PipRequirement`.
+   - This will require to create conda packages for bears that have different
+     installation instructions for each distribution. This is done by making `bld.bat`
+     file that has installation instruction for Windows and similarly a `build.sh` file
+     that has installation instruction for Linux.
+3. After installing the bear we proceed to installing its requirements. We call
+   methods from the requirement classes to do this.
+   - Add new Requirement classes to support different dependencies from various linters.
+   - The DistributionRequirement class needs to be split into different classes for each
+     package manager and call its method to introduce more consistency.
+4. We then proceed to install bear dependencies such as for `ClangCloneDetectionBear` by
+   running `install_bears` function for each dependency bear.
+
+### Conda Packaging
+
+1. Create a new repository called `coala-bear packages`.
+2. This repository will have a folder called `conda packages` that contains conda packages for
+   all the bears that require different installation instructions on each distribution. Each
+   conda package requires 3 files i.e. `meta.yaml`, `build.sh`, `bld.bat`.
+3. Move the `generate_package` to this repository and allow it to generate conda packages only for
+   the bears that are in the folder `conda packages`.
+
+### Process for upgrading bears in cib
+
+1. Check the bear name is installed using `CondaRequirement` or `PipRequirement`.
+2. Iterate through `Requirement` instances to make sure that the required version of the
+   linter is installed or upgrade the version if required.
+
+### Split `DistributionRequirement` for each distribution
+
+This is the list of all distributions <http://distro.readthedocs.io/en/latest/#distro.id>
+available from `distro` package (excluding Darwin and Windows). Supporting all these
+platforms would be too broad for this project. Therefore, the following platforms listed
+below have been chosen to be completed as a part of this project.
+
+| Operating System | Tools   |
+| ---------------- | ------- |
+| Debian/Ubuntu    | apt-get |
+| CentOS           | yum     |
+| Fedora           | dnf     |
+| Arch Linux       | pacman  |
+| openSUSE/SUSE    | zypper  |
+| Gentoo Linux     | portage |
+| Darwin           | brew    |
+
+1. Create a class `PlatformRequirement` representing a general requirement for a given
+   distribution. This would be similiar to `PackageRequirement` but it will have a method
+   to check whether the user is on the correct platform.
+2. Then create classes such as `AptRequirement`, `YumRequirement` etc. that will inherit
+   from the `PlatformRequirement` class.
+3. Deprecate the `DistributionRequirement` class.
+
+### Ansible for downloading bear and dependencies
+
+1. Write a script in python to create a playbook that will have the installation instruction
```

Comment:
we should have individual playbooks for each bear, as most users only want a few bears.

a playbook for all bears is also very helpful, and could be used on our docker image

Reply:
then the users would have to download each playbook and run `ansible-playbook bear_name.yml` that would install the bear. 

But also my idea with the playbook for all bears was to have tags so that the user would run `ansible-playbook bears.yml --tags "EslintBear, VintBear"` to run the required tasks for installing the specified bears.

However, it should be easy to create separate playbooks given the script.